pkg/operator/ceph/reporting/reporting.go

```diff
@@ -0,0 +1,124 @@
+/*
+Copyright 2021 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Reporting focuses on reporting Events, Status Conditions, and the like to users.
+package reporting
+
+import (
+	"fmt"
+
+	"github.com/coreos/pkg/capnslog"
+	"github.com/pkg/errors"
+	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	"github.com/rook/rook/pkg/operator/k8sutil"
+	"github.com/rook/rook/pkg/util/dependents"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// ReportReconcileResult will report the result of an object's reconcile in 2 ways:
+// 1. to the given logger
+// 2. as an event on the object (via the given event recorder)
+// The results of the object's reconcile should include the object, the reconcile repsonse, and the
+// error returned by the reconcile.
+// The function is designed to return the appropriate values needed for the controller-runtime
+// framework's Reconcile() method.
+func ReportReconcileResult(logger *capnslog.PackageLogger, recorder *k8sutil.EventReporter,
+	obj client.Object, reconcileResponse reconcile.Result, err error,
+) (reconcile.Result, error) {
+	kind := obj.GetObjectKind().GroupVersionKind().Kind
+	nsName := fmt.Sprintf("%s/%s", obj.GetNamespace(), obj.GetName())
+
+	if err != nil {
+		// 1. log
+		logger.Errorf("failed to reconcile %s %q. %v", kind, nsName, err)
+
+		// 2. event
+		recorder.ReportIfNotPresent(obj, corev1.EventTypeWarning, string(cephv1.ReconcileFailed), err.Error())
+
+		if !reconcileResponse.IsZero() {
+			// The framework will requeue immediately if there is an error. If we get an error with
+			// a non-empty reconcile response, just return the response with the error now logged as
+			// an event so that the framework can pause before the next reconcile per the response's
+			// intent.
+			return reconcileResponse, nil
+		}
+	} else {
+		successMsg := fmt.Sprintf("successfully configured %s %q", kind, nsName)
+
+		// 1. log
+		logger.Debug(successMsg)
```

Comment:
Why not just use `Debugf()`?

Reply:
`successMsg` gets reused multiple times in this function.